src/circuits/atomicV3.ts

```diff
@@ -0,0 +1,144 @@
+import { PubSignalsVerifier, VerifyOpts } from '@lib/circuits/registry';
+import { checkQueryRequest, ClaimOutputs, Query } from '@lib/circuits/query';
+import { Resolvers } from '@lib/state/resolver';
+import { IDOwnershipPubSignals } from '@lib/circuits/ownershipVerifier';
+import { checkIssuerNonRevState, checkUserState, getResolverByID } from '@lib/circuits/common';
+import { DID, getDateFromUnixTimestamp } from '@iden3/js-iden3-core';
+import { DocumentLoader } from '@iden3/js-jsonld-merklization';
+import {
+  AtomicQueryV3PubSignals,
+  BaseConfig,
+  byteEncoder,
+  JSONObject,
+  ProofType
+} from '@0xpolygonid/js-sdk';
+
+const valuesSize = 64;
+const defaultProofVerifyOpts = 1 * 60 * 60 * 1000; // 1 hour
+
+export class AtomicQueryV3PubSignalsVerifier
+  extends IDOwnershipPubSignals
+  implements PubSignalsVerifier
+{
+  pubSignals = new AtomicQueryV3PubSignals();
+
+  constructor(pubSignals: string[]) {
+    super();
+    this.pubSignals = this.pubSignals.pubSignalsUnmarshal(
+      byteEncoder.encode(JSON.stringify(pubSignals))
+    );
+
+    this.userId = this.pubSignals.userID;
+    this.challenge = this.pubSignals.requestID;
+  }
+
+  async verifyQuery(
+    query: Query,
+    schemaLoader?: DocumentLoader,
+    verifiablePresentation?: JSON,
+    opts?: VerifyOpts,
+    params?: JSONObject
+  ): Promise<BaseConfig> {
+    const outs: ClaimOutputs = {
+      issuerId: this.pubSignals.issuerID,
+      schemaHash: this.pubSignals.claimSchema,
+      slotIndex: this.pubSignals.slotIndex,
+      operator: this.pubSignals.operator,
+      value: this.pubSignals.value,
+      timestamp: this.pubSignals.timestamp,
+      merklized: this.pubSignals.merklized,
+      claimPathKey: this.pubSignals.claimPathKey,
+      claimPathNotExists: this.pubSignals.claimPathNotExists,
+      valueArraySize: valuesSize,
+      isRevocationChecked: this.pubSignals.isRevocationChecked
+    };
+    await checkQueryRequest(query, outs, schemaLoader, verifiablePresentation, opts);
+
+    const { proofType, verifierID, nullifier, nullifierSessionID, linkID } = this.pubSignals;
+
+    switch (query.proofType) {
+      case ProofType.BJJSignature:
+        if (proofType !== 1) {
+          throw new Error('wrong proof type for BJJSignature');
+        }
+        break;
+      case ProofType.Iden3SparseMerkleTreeProof:
+        if (proofType !== 2) {
+          throw new Error('wrong proof type for Iden3SparseMerkleTreeProof');
+        }
+        break;
+      default:
+        throw new Error('invalid proof type');
+    }
+
+    const nullifierSessionIDparam = params?.nullifierSessionId;
+
+    if (nullifierSessionIDparam) {
+      if (nullifier && BigInt(nullifier) !== 0n) {
```

Comment:
should we throw if `nullifierSessionIDparam`exists and `nullifier` is 0 ? 

maybe add inside `if (nullifierSessionIDparam)` 
`if (!nullifier || BigInt(nullifier) === 0n) throw new Error('msg')`

Reply:
I see that we have check ` if (nullifierSessionID !== nSessionId) {` 
but we won't get thereif `! nullifier `

Reply:
should we throw if nullifierSessionIDparamexists and nullifier is 0 - yes .